common/browser_context.go

```diff
@@ -546,6 +547,131 @@ func (b *BrowserContext) Cookies(urls ...string) ([]*api.Cookie, error) { //noli
 			SameSite: api.CookieSameSite(c.SameSite),
 		}
 	}
+	// filter cookies by the provided URLs.
+	cookies, err = filterCookies(cookies, urls...)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving cookies: %w", err)
+	}
+	if len(cookies) == 0 {
+		return nil, nil
+	}
 
 	return cookies, nil
 }
+
+// filterCookies filters the given cookies based on URLs.
+// If an error occurs while parsing the cookie URLs, the error is returned.
+func filterCookies(cookies []*api.Cookie, urls ...string) ([]*api.Cookie, error) {
+	if len(urls) == 0 || len(cookies) == 0 {
+		return cookies, nil
+	}
+
+	purls, err := parseURLs(urls...)
+	if err != nil {
+		return nil, fmt.Errorf("filtering by URL: %w", err)
+	}
+
+	// the following algorithm is like a sorting algorithm,
+	// but instead of sorting, it filters the cookies slice
+	// in place, without allocating a new slice. this is
+	// done to avoid unnecessary allocations and copying
+	// of data.
+	//
+	// n is used to remember the last cookie that should be
+	// kept in the cookies slice. all cookies before n should
+	// be kept, all cookies after n should be removed. it
+	// constantly shifts cookies to be kept to the left in the
+	// slice, overwriting cookies that should be removed.
+	//
+	// if a cookie should not be kept, it will be overwritten
+	// by the next cookie that should be kept. if no cookies
+	// should be kept, a nil slice is returned. otherwise,
+	// the slice is truncated to the last cookie that should
+	// be kept.
+
+	var n int
+
+	for _, c := range cookies {
+		var keep bool
+
+		for _, uri := range purls {
+			if shouldFilterCookie(c, uri) {
+				keep = true
+				break
+			}
+		}
+		if !keep {
+			continue
+		}
+		cookies[n] = c
+		n++
+	}
+	// if no cookies should be kept, return nil instead of
+	// an empty slice to conform with the API error behavior.
+	// also makes tests concise.
+	if n == 0 {
+		return nil, nil
+	}
+
+	// remove all cookies after the last cookie that should be kept.
+	return cookies[:n], nil
+}
+
+// shouldFilterCookie determines whether a cookie should be kept,
+// based on its compatibility with a specific URL.
+// Returns true if the cookie should be kept, false otherwise.
+func shouldFilterCookie(c *api.Cookie, uri *url.URL) bool {
+	// Ensure consistent domain formatting for easier comparison.
+	// A leading dot means the cookie is valid across subdomains.
+	// For example, if the domain is example.com, then adding a
+	// dot turns it into .example.com, making the cookie valid
+	// for sub.example.com, another.example.com, etc.
+	domain := c.Domain
+	if !strings.HasPrefix(domain, ".") {
+		domain = "." + domain
+	}
+	// Confirm that the cookie's domain is a suffix of the URL's
+	// hostname, emulating how a browser would scope cookies to
+	// specific domains.
+	if !strings.HasSuffix("."+uri.Hostname(), domain) {
```

Comment:
Possibly not a big deal, but shouldn't this be the other way around:

```suggestion
	if !strings.HasSuffix(domain, "."+uri.Hostname()) {
```

So now I would read that as _if the cookie's domain doesn't match the expected filter uri hostname_... WDYT?